Kernel/FileSystem/FATFS/Inode.cpp

```diff
@@ -55,26 +55,116 @@ ErrorOr<Vector<BlockBasedFileSystem::BlockIndex>> FATInode::compute_block_list()
     auto fat_sector = TRY(KBuffer::try_create_with_size("FATFS: FAT read buffer"sv, fs().m_device_block_size));
     auto fat_sector_buffer = UserOrKernelBuffer::for_kernel_buffer(fat_sector->data());
 
-    while (cluster < no_more_clusters) {
+    while (cluster < no_more_clusters()) {
         dbgln_if(FAT_DEBUG, "FATFS: Appending cluster {} to inode {}'s cluster chain", cluster, index());
 
-        BlockBasedFileSystem::BlockIndex first_block = fs().first_block_of_cluster(cluster);
-        for (u8 i = 0; i < fs().boot_record()->sectors_per_cluster; i++)
-            block_list.append(BlockBasedFileSystem::BlockIndex { first_block.value() + i });
+        Tuple<BlockBasedFileSystem::BlockIndex, size_t> first_block_and_length = fs().first_block_of_cluster(cluster);
+        for (u8 i = 0; i < first_block_and_length.get<1>(); i++)
+            block_list.append(BlockBasedFileSystem::BlockIndex { first_block_and_length.get<0>().value() + i });
+
+        // Clusters 0 and 1 are reserved in the FAT, and their entries in the FAT will
+        // not point to another valid cluster in the chain (Cluster 0 typically holds
+        // the "FAT ID" field with some flags, Cluster 1 should be the end of chain
+        // marker).
+        // Internally, we use `cluster == 0` to represent the root directory Inode,
+        // which is a signal to read the root directory region blocks on FAT12/16
+        // file systems. (`fs().first_block_of_cluster` will return the appropriate
+        // block/sectors to read given cluster == 0).
+        // Therefore, we read one set of sectors for these invalud cluster numbers,
+        // and then terminate the loop becuase the FAT entry at `cluster` for these
+        // values does not represent the next step in the chain (because there is
+        // nothing else to read).
+        if (cluster <= 1) {
+            break;
+        }
 
-        u32 fat_offset = cluster * sizeof(u32);
-        u32 fat_sector_index = fs().boot_record()->reserved_sector_count + (fat_offset / fs().m_device_block_size);
+        u32 fat_offset = fat_offset_for_cluster(cluster);
+        u32 fat_sector_index = fs().m_parameter_block->block->reserved_sector_count + (fat_offset / fs().m_device_block_size);
         u32 entry_offset = fat_offset % fs().m_device_block_size;
 
         TRY(fs().raw_read(fat_sector_index, fat_sector_buffer));
 
-        cluster = *reinterpret_cast<u32*>(&fat_sector->data()[entry_offset]);
-        cluster &= cluster_number_mask;
+        // Look up the next cluster to read, or read End of Chain marker from table.
+        cluster = cluster_number(*fat_sector, cluster, entry_offset);
     }
 
     return block_list;
 }
 
+u32 FATInode::no_more_clusters() const
```

Comment:
Please review <https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#names>

This method should at the very least have a verb at the start

`is_final_cluster()` or `get_cluster_terminator` or some such.

Reply:
Renamed to `get_end_of_chain_marker()`.